llarp/messages/relay_status.cpp

```diff
@@ -0,0 +1,295 @@
+      llarp_threadpool_queue_job(router->pathContext().Worker(), func);
+    }
+
+  };
+
+
+  LR_StatusMessage::~LR_StatusMessage()
+  {
+  }
+
+  bool
+  LR_StatusMessage::DecodeKey(const llarp_buffer_t& key, llarp_buffer_t* buf)
+  {
+    bool read = false;
+    if(key == "c")
+    {
+      return BEncodeReadArray(frames, buf);
+    }
+    else if(key == "p")
+    {
+      if (!BEncodeMaybeReadDictEntry("p", pathid, read, key, buf))
+      {
+        return false;
+      }
+    }
+    else if(key == "s")
+    {
+      if (!BEncodeMaybeReadDictInt("s", status, read, key, buf))
+      {
+        return false;
+      }
+    }
+    else if(key == "v")
+    {
+      if(!BEncodeMaybeReadVersion("v", version, LLARP_PROTO_VERSION, read, key,
+                                  buf))
+      {
+        return false;
+      }
+    }
+
+    return read;
+  }
+
+  void
+  LR_StatusMessage::Clear()
+  {
+    frames[0].Clear();
+    frames[1].Clear();
+    frames[2].Clear();
+    frames[3].Clear();
+    frames[4].Clear();
+    frames[5].Clear();
+    frames[6].Clear();
+    frames[7].Clear();
+  }
+
+  bool
+  LR_StatusMessage::BEncode(llarp_buffer_t* buf) const
+  {
+    if(!bencode_start_dict(buf))
+      return false;
+    // msg type
+    if(!BEncodeWriteDictMsgType(buf, "a", "s"))
+      return false;
+    // frames
+    if(!BEncodeWriteDictArray("c", frames, buf))
+      return false;
+    // path id
+    if(!BEncodeWriteDictEntry("p", pathid, buf))
+      return false;
+    // status (for now, only success bit is relevant)
+    if(!BEncodeWriteDictInt("s", status, buf))
+      return false;
+    // version
+    if(!bencode_write_version_entry(buf))
+      return false;
+
+    return bencode_end(buf);
+  }
+
+  bool
+  LR_StatusMessage::HandleMessage(AbstractRouter* router) const
+  {
+    llarp::LogDebug("Received LR_Status message from (", session->GetPubKey(), ")");
+    if(frames.size() != path::max_len)
+    {
+      llarp::LogError("LRSM invalid number of records, ", frames.size(),
+                      "!=", path::max_len);
+      return false;
+    }
+
+    auto path = router->pathContext().GetByUpstream(session->GetPubKey(), pathid);
+    if (!path)
+    {
+      llarp::LogWarn("unhandled LR_Status message: no associated IHopHandler found");
+      return false;
+    }
+
+    auto handler = std::make_shared< LRSM_AsyncHandler >(
+        frames, pathid, status, path, router);
+
+    handler->queue_handle();
+
+    return true;
+  }
+
+  void LR_StatusMessage::SetDummyFrames()
+  {
+    //TODO
+    return;
+  }
+
+  // call this from a worker thread
+  bool
+  LR_StatusMessage::CreateAndSend(AbstractRouter *router, const PathID_t pathid,
+                                  const RouterID nextHop, const SharedSecret pathKey,
+                                  uint64_t status)
+  {
+    typedef std::shared_ptr< LR_StatusMessage > LRSM_ptr;
+    LRSM_ptr message = std::make_shared< LR_StatusMessage >();
```

Comment:
remove the typedef (which should be [`using`](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rt-using), and then:

`auto message = std::make_shared< LR_StatusMessage >();`

Reply:
With auto I may even be able to omit the `using` bit.  Will revisit.

Reply:
addressed